src/org/labkey/remoteapi/query/SaveRowsApiCommand.java

```diff
@@ -0,0 +1,288 @@
+    {
+        _extraContext = extraContext;
+        return this;
+    }
+
+    public SaveRowsApiCommand addCommand(Command... commands)
+    {
+        for (Command command : commands)
+        {
+            if (command != null)
+                _commands.add(command);
+        }
+
+        return this;
+    }
+
+    public List<Command> getCommands()
+    {
+        return _commands;
+    }
+
+    public Boolean isTransacted()
+    {
+        return _transacted;
+    }
+
+    public SaveRowsApiCommand setTransacted(Boolean transacted)
+    {
+        _transacted = transacted;
+        return this;
+    }
+
+    public Boolean isValidateOnly()
+    {
+        return _validateOnly;
+    }
+
+    public SaveRowsApiCommand setValidateOnly(Boolean validateOnly)
+    {
+        _validateOnly = validateOnly;
+        return this;
+    }
+
+    @Override
+    public JSONObject getJsonObject()
+    {
+        JSONObject json = new JSONObject();
+
+        List<JSONObject> commands = new ArrayList<>();
+        for (Command command : getCommands())
+            commands.add(command.getJsonObject());
+        json.put("commands", commands);
+
+        if (getExtraContext() != null && !getExtraContext().isEmpty())
+            json.put("extraContext", getExtraContext());
+
+        if (isTransacted() != null)
+            json.put("transacted", isTransacted());
+
+        if (isValidateOnly() != null)
+            json.put("validateOnly", isValidateOnly());
+
+        return json;
+    }
+
+    @Override
+    protected SaveRowsApiResponse createResponse(String text, int status, String contentType, JSONObject json)
+    {
+        return new SaveRowsApiResponse(text, status, contentType, json);
+    }
+
+    public enum CommandType
+    {
+        Insert,
+        Update,
+        Delete
+    }
+
+    public static class Command
+    {
+        SaveRowsCommand.AuditBehavior _auditBehavior;
+        String _auditUserComment;
+        final CommandType _commandType;
+        String _containerPath;
+        Map<String, Object> _extraContext;
+        List<Map<String, Object>> _rows;
+        final String _queryName;
+        final String _schemaName;
+        Boolean _skipReselectRows;
+
+        public Command(CommandType commandType, String schemaName, String queryName, List<Map<String, Object>> rows)
+        {
+            assert null != commandType;
+            assert null != schemaName && !schemaName.isEmpty();
+            assert null != queryName && !queryName.isEmpty();
+
+            _commandType = commandType;
+            _schemaName = schemaName;
+            _queryName = queryName;
+            _rows = rows;
+        }
+
+        public JSONObject getJsonObject()
+        {
+            JSONObject json = new JSONObject();
+
+            json.put("command", getCommandType().name().toLowerCase());
+            json.put("schemaName", getSchemaName());
+            json.put("queryName", getQueryName());
+            json.put("rows", getRows());
```

Comment:
The counterpart in `SaveRowsCommand` has notes about the need for particular processing for maps and dates. Is that not needed here or is it not needed there?